### ISSUE TITLE
Reach go/no-go decision for Cranelift ZK Backend

### ISSUE THREAD

**Issue by akashin** [near/nearcore#9514]
Our [current approach](
https://github.com/near/wasmtime/tree/zkasm-hackaton) to building a ZK prover for WASM programs is based on the [Cranelift](https://cranelift.dev/) tech stack. Concretely, we implement [a new backend/ISA](https://github.com/near/wasmtime/tree/zkasm-hackaton/cranelift/codegen/src/isa/zkasm) that can be targeted by the Cranelift compiler which allows us to build ZK circuits through the following pipeline: WASM -> Cranelift IR -> ZK ASM
This backend was started as a [hackathon](https://hackmd.io/8faPpmGQTrmdr1mZau6XYQ?view) project with @Ekleog  and @nagisa.

We want to understand whether this tech stack is the right choice for ZK WASM project for the [next 6 months](https://github.com/near/nearcore/issues/9513) and reach the go/no-go decision by the end of September 2023.

To do this, we need to:

```[tasklist]
- [ ] List and evaluate alternative solutions that we can consider
- [ ] Write down the list of hard constraints of the project and understand whether Cranelift will satisfy them
- [ ] Get advice from [Cranelift developers](https://docs.wasmtime.dev/contributing.html) by describing our problem, and understanding what it would take to upstream our backend to Cranelift core project
```

Some prior art:
- https://near.zulipchat.com/#narrow/stream/295306-pagoda.2Fcontract-runtime/topic/llvm.20backend.20for.20zk
- https://www.risczero.com/

**Comment by akashin** [comment]
I'm working on a [document](https://hackmd.io/@ke2spS4iQmu1IxfGt3-qoQ/Sks1aOlk6) that I plan to share with Cranelift developers to get their advice and ideally set up a meeting to discuss it in more detail.

**Comment by akashin** [comment]
@nagisa started the conversation with Cranelift folks: https://bytecodealliance.zulipchat.com/#narrow/stream/217117-cranelift/topic/zkASM.20backend and we aim to attend their weekly meeting with a short presentation about ZK WASM.

We also have a better understanding of what it takes to merge code into wasmtime upstream: https://docs.wasmtime.dev/stability-tiers.html#tier-3---not-production-ready
In short, we will need to:
* Make our code aligned with abstractions used in Cranelift (e.g. bytecode generation)
* Get the code to high quality and testing standard
* Designate a maintainer for a ZK ASM backend who will respond to related wasmtime interface changes

These goals are already on our roadmap, so it seems upsteaming our backend would be well aligned with our efforts.
